### PR TITLE
Implement `set_attribute` & `set_class_attribute` nonfungibles trait functions for `Uniques` pallet

### PR DESCRIPTION
**Pull request by iorveth**
Implements missing `set_attribute`, `set_class_attribute` `nonfungibles` `Mutate` trait functions for `uniques` pallet
Relates to https://github.com/paritytech/substrate/issues/9208

### CODE REVIEW

File: frame/uniques/src/functions.rs [hunk 1]
```diff
@@ -112,4 +112,47 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+        class_details.total_deposit.saturating_reduce(old_deposit);
+        let mut deposit = Zero::zero();
+        if !class_details.free_holding && maybe_check_owner.is_some() {
+            deposit = T::DepositPerByte::get()
+                .saturating_mul(((key.len() + value.len()) as u32).into())
+                .saturating_add(T::AttributeDepositBase::get());
+        }
+        class_details.total_deposit.saturating_accrue(deposit);
+        if deposit > old_deposit {
+            T::Currency::reserve(&class_details.owner, deposit - old_deposit)?;
```
Comment (apopiak) [inline frame/uniques/src/functions.rs:147]:
```suggestion
            T::Currency::reserve(&class_details.owner, deposit.saturating_sub(old_deposit))?;
```

File: frame/uniques/src/functions.rs [hunk 2]
```diff
@@ -112,4 +112,47 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+        } else if deposit < old_deposit {
+            T::Currency::unreserve(&class_details.owner, old_deposit - deposit);
```
Comment (apopiak) [inline frame/uniques/src/functions.rs:149]:
```suggestion
            T::Currency::unreserve(&class_details.owner, old_deposit.saturating_sub(deposit));
```

File: frame/uniques/src/functions.rs [hunk 3]
```diff
@@ -112,4 +112,47 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+        let attribute = Attribute::<T, I>::get((class, maybe_instance, &key));
+        if attribute.is_none() {
+            class_details.attributes.saturating_inc();
+        }
+        let old_deposit = attribute.map_or(Zero::zero(), |m| m.1);
+        class_details.total_deposit.saturating_reduce(old_deposit);
+        let mut deposit = Zero::zero();
+        if !class_details.free_holding && maybe_check_owner.is_some() {
+            deposit = T::DepositPerByte::get()
+                .saturating_mul(((key.len() + value.len()) as u32).into())
```
Comment (apopiak) [inline frame/uniques/src/functions.rs:142]:
```suggestion
                .saturating_mul((key.len().saturating_add(value.len()) as u32).into())
```

File: frame/uniques/src/impl_nonfungibles.rs [hunk 4]
```diff
@@ -95,6 +95,43 @@ impl<T: Config<I>, I: 'static> Mutate<<T as SystemConfig>::AccountId> for Pallet
...
+
+        Self::do_set_attribute(
+            class.clone(),
+            Some(instance.clone()),
+            &None,
+            bounded_key,
+            bounded_value,
+            |_| Ok(()),
+        )
+    }
```
[the remark this answers is not in the document]
Comment (iorveth) [reply frame/uniques/src/impl_nonfungibles.rs:116]: Done https://github.com/paritytech/substrate/pull/9206/commits/eb6924d7ebd82ed63a131c684ab00907a682246e

File: frame/uniques/src/functions.rs (hunk 3 above)
Comment (iorveth) [reply frame/uniques/src/functions.rs]: Thank you for the suggestion, done here https://github.com/paritytech/substrate/pull/9206/commits/6fabf252387b41d081b849eab6acff64b060d253

File: frame/uniques/src/functions.rs (hunk 2 above)
Comment (iorveth) [reply frame/uniques/src/functions.rs]: Thank you for the suggestion, done here https://github.com/paritytech/substrate/pull/9206/commits/6fabf252387b41d081b849eab6acff64b060d253

File: frame/uniques/src/functions.rs (hunk 1 above)
Comment (iorveth) [reply frame/uniques/src/functions.rs]: Thank you for the suggestion, done here https://github.com/paritytech/substrate/pull/9206/commits/6fabf252387b41d081b849eab6acff64b060d253

**Comment by shawntabrizi** [comment]
This looks okay, but I suggest we do not increase the expectations of the `Mutate` trait, and instead introduce a new trait which inherits from `Mutate` called `MutateWithMetadata` or something.

Remember that we should not assume all (or even most) NFT systems will require some on-chain metadata.

File: frame/uniques/src/functions.rs (hunk 1 above)
Comment (shawntabrizi) [reply frame/uniques/src/functions.rs]: This is NOT needed, and actually probably some additional overhead.

Note the check just 1 line above: `deposit > old_deposit`

This subtraction can never panic

File: frame/uniques/src/functions.rs (hunk 2 above)
Comment (shawntabrizi) [reply frame/uniques/src/functions.rs]: undo

File: frame/uniques/src/functions.rs (hunk 3 above)
Comment (shawntabrizi) [reply frame/uniques/src/functions.rs]: this one i think is okay, but would want you to look closer before actually making this change

and then a test

**Pull request closed without merging**